jerry-core/ecma/builtin-objects/ecma-builtin-helpers.c

```diff
@@ -532,73 +532,164 @@ ecma_builtin_helper_string_index_normalize (ecma_number_t index, /**< index */
+  if (ECMA_IS_VALUE_ERROR (ecma_op_check_object_coercible (this_arg)))
+  {
+    return ECMA_VALUE_ERROR;
+  }
 
   /* 2 */
-  ECMA_TRY_CATCH (to_str_val,
-                  ecma_op_to_string (this_arg),
-                  ret_value);
-
-  /* 3 */
-  ECMA_TRY_CATCH (search_str_val,
-                  ecma_op_to_string (arg1),
-                  ret_value);
+  ecma_value_t to_str_val = ecma_op_to_string (this_arg);
 
-  /* 4 */
-  ECMA_OP_TO_NUMBER_TRY_CATCH (pos_num,
-                               arg2,
-                               ret_value);
+  if (ECMA_IS_VALUE_ERROR (to_str_val))
+  {
+    return to_str_val;
+  }
 
-  /* 5 (indexOf) -- 6 (lastIndexOf) */
+  /* 5 (indexOf), 6 (lastIndexOf), 11 (startsWith, includes) */
   ecma_string_t *original_str_p = ecma_get_string_from_value (to_str_val);
   const ecma_length_t original_len = ecma_string_get_length (original_str_p);
 
-  /* 4b, 6 (indexOf) - 4b, 5, 7 (lastIndexOf) */
-  ecma_length_t start = ecma_builtin_helper_string_index_normalize (pos_num, original_len, first_index);
+  #if ENABLED (JERRY_ES2015_BUILTIN)
+  /* 4, 6 (startsWith, includes, endsWith) */
+  if ((mode > ECMA_STRING_LAST_INDEX_OF)
+      && (ecma_is_value_object (arg1)
+      && ecma_object_class_is (ecma_get_object_from_value (arg1), LIT_MAGIC_STRING_REGEXP_UL)))
+  {
+    ecma_deref_ecma_string (original_str_p);
+    return ecma_raise_type_error (ECMA_ERR_MSG ("Search string can't be of type: RegExp"));
+  }
+  #endif /* ENABLED (JERRY_ES2015_BUILTIN) */
 
-  /* 7 (indexOf) -- 8 (lastIndexOf) */
+  /* 3 */
+  ecma_value_t search_str_val = ecma_op_to_string (arg1);
+
+  if (ECMA_IS_VALUE_ERROR (search_str_val))
+  {
+    ecma_deref_ecma_string (original_str_p);
+    return search_str_val;
+  }
+
+  /* 7, 8 */
   ecma_string_t *search_str_p = ecma_get_string_from_value (search_str_val);
 
+  /* 4 (indexOf, lastIndexOf), 9 (startsWith, includes), 10 (endsWith) */
+  ecma_number_t pos_num;
+  ecma_value_t ret_value  = ecma_get_number (arg2, &pos_num);
+
+  /* 10 (startsWith, includes), 11 (endsWith) */
+  if (ECMA_IS_VALUE_ERROR (ret_value))
+  {
+    ecma_deref_ecma_string (original_str_p);
+    ecma_deref_ecma_string (search_str_p);
+    return ret_value;
+  }
+
+  /* 4b, 6 (indexOf) - 4b, 5, 7 (lastIndexOf) */
+  ecma_length_t start = ecma_builtin_helper_string_index_normalize (pos_num, original_len, first_index);
+
   ecma_number_t ret_num = ECMA_NUMBER_MINUS_ONE;
 
-  /* 8 (indexOf) -- 9 (lastIndexOf) */
   ecma_length_t index_of = 0;
-  if (ecma_builtin_helper_string_find_index (original_str_p, search_str_p, first_index, start, &index_of))
+
+  ret_value = ECMA_VALUE_FALSE;
+
+  switch (mode)
   {
-    ret_num = ((ecma_number_t) index_of);
-  }
+    #if ENABLED (JERRY_ES2015_BUILTIN)
+    case ECMA_STRING_STARTS_WITH:
+    {
+      if (pos_num + start > original_len)
+      {
+        break;
+      }
+      if (ecma_builtin_helper_string_find_index (original_str_p, search_str_p, first_index, start, &index_of))
```

Comment:
`first_index` should be true all the way, except if it's `lastIndex`. Removed the variable from the places where it was `true` all the way and left it where it could be both. Looks nicer indeed. :smile_cat: